physical/etcd3.go

```diff
@@ -0,0 +1,285 @@
+	}
+
+	etcd, err := clientv3.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	sync, ok := conf["sync"]
+	if !ok {
+		sync = "yes"
+	}
+	switch sync {
+	case "yes", "true", "y", "1":
+		ctx, cancel := context.WithTimeout(context.Background(), client.DefaultRequestTimeout)
+		err := etcd.Sync(ctx)
+		cancel()
+		if err != nil {
+			return nil, err
+		}
+	case "no", "false", "n", "0":
+	default:
+		return nil, fmt.Errorf("value of 'sync' could not be understood")
+	}
+
+	return &Etcd3Backend{
+		path:       path,
+		etcd:       etcd,
+		permitPool: NewPermitPool(DefaultParallelOperations),
+		logger:     logger,
+		haEnabled:  haEnabledBool,
+	}, nil
+}
+
+func (c *Etcd3Backend) Put(entry *Entry) error {
+	defer metrics.MeasureSince([]string{"etcd", "put"}, time.Now())
+
+	c.permitPool.Acquire()
+	defer c.permitPool.Release()
+
+	_, err := c.etcd.Put(context.Background(), path.Join(c.path, entry.Key), string(entry.Value))
+	return err
+}
+
+func (c *Etcd3Backend) Get(key string) (*Entry, error) {
+	defer metrics.MeasureSince([]string{"etcd", "get"}, time.Now())
+
+	c.permitPool.Acquire()
+	defer c.permitPool.Release()
+
+	resp, err := c.etcd.Get(context.Background(), path.Join(c.path, key))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.Kvs) == 0 {
+		return nil, nil
+	}
+	return &Entry{
+		Key:   key,
+		Value: resp.Kvs[0].Value,
+	}, nil
+}
+
+func (c *Etcd3Backend) Delete(key string) error {
+	defer metrics.MeasureSince([]string{"etcd", "delete"}, time.Now())
+
+	c.permitPool.Acquire()
+	defer c.permitPool.Release()
+
+	_, err := c.etcd.Delete(context.Background(), path.Join(c.path, key))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Etcd3Backend) List(prefix string) ([]string, error) {
+	defer metrics.MeasureSince([]string{"etcd", "list"}, time.Now())
+
+	c.permitPool.Acquire()
+	defer c.permitPool.Release()
+
+	prefix = path.Join(c.path, prefix)
+	resp, err := c.etcd.Get(context.Background(), prefix, clientv3.WithPrefix())
+	if err != nil {
+		return nil, err
+	}
+
+	keys := []string{}
+	for _, kv := range resp.Kvs {
+		key := strings.TrimPrefix(string(kv.Key), prefix)
+		key = strings.TrimPrefix(key, "/")
+
+		if len(key) == 0 {
+			continue
+		}
+
+		if i := strings.Index(key, "/"); i == -1 {
+			keys = append(keys, key)
+		} else if i != -1 {
+			keys = appendIfMissing(keys, key[:i+1])
+		}
+	}
+	return keys, nil
+}
+
+func (e *Etcd3Backend) HAEnabled() bool { return e.haEnabled }
+
+// EtcdLock emplements a lock using and etcd backend.
+type Etcd3Lock struct {
+	lock sync.Mutex
+	held bool
+
+	etcdSession *concurrency.Session
+	etcdMu      *concurrency.Mutex
+
+	prefix string
+	value  string
+
+	etcd *clientv3.Client
+}
+
+// Lock is used for mutual exclusion based on the given key.
+func (c *Etcd3Backend) LockWith(key, value string) (Lock, error) {
+	session, err := concurrency.NewSession(c.etcd)
+	if err != nil {
+		return nil, err
+	}
+
+	p := path.Join(c.path, "/__lock__/", key)
```

Comment:
is it a requirement? the problem is that our lock implementation does not support lock on a real key. it actually locks on a virtual key.

if you specify a lock with key `foo`, it will create a queue under prefix `foo` like:

`foo/1`
`foo/2`

So if you have a client read the value of key `foo`, it wont return the value of `foo/1`. I am not sure if this is OK.